File: _outils/fiches/paginer.py
#!/usr/bin/env python3
"""
paginer.py — propose la découpe en pages d'une fiche, à la mesure.

POURQUOI — `mesurer_pages.py` dit *après coup* où le contenu s'arrête sur
chaque page du PDF. Il ne dit pas **où couper**. Tant qu'on déplaçait les
appels `feuille()` à la main, chaque essai coûtait une génération, un export
et une lecture : sur T1-C2, la première découpe laissait 607 mm de creux et
une page qui débordait. Cet outil mesure chaque bloc **isolément**, puis
calcule les coupes qui équilibrent les pages. Le creux est tombé à 122 mm.

CE QU'IL SUPPOSE — que le fichier de chapitre expose deux fonctions :

    codes_qr(muet=False) -> dict des QR
    blocs(src, qr)       -> [(partie | None, html), ...]  dans l'ordre

et une constante `COUPES`. `fiche_t1c2.py` est le modèle. Les fiches écrites
avant (T3-C1) construisent leurs pages à la main : elles marchent toujours,
elles ne sont simplement pas paginables par cet outil.

USAGE
    python3 paginer.py t1c2            # propose 10, 11, 12, 13, 14 pages
    python3 paginer.py t1c2 12         # impose 12 pages
    python3 paginer.py t1c2 12 260     # ... avec un autre budget par page

Puis recopier la ligne `COUPES = [...]` dans le fichier de chapitre.

🔴 CET OUTIL PROPOSE, IL NE TRANCHE PAS. Le budget de 267 mm est une valeur
empirique, et la somme des blocs sous-estime la page de 0 à 21 mm selon le
nombre d'en-têtes qu'elle porte (mesuré sur les douze pages de T1-C2). Ce
qui fait foi reste :

    node exporter-fiches.mjs <code>
    python3 _outils/fiches/mesurer_pages.py assets/pdf/pc/fiches/fiche-2nde-<code>.pdf

Marche à suivre : proposer ici, générer, exporter, mesurer — et si une page
déborde, retirer une ou deux lignes au cadre le plus généreux de cette page
plutôt que de tout repaginer. C'est exactement ce qu'a demandé T1-C2 : une
seule page sur douze débordait, l'exercice 10 est passé de 10 à 8 lignes.
"""

import importlib, subprocess, sys, tempfile
from pathlib import Path

sys.stdout.reconfigure(encoding="utf-8", errors="replace")
ICI = Path(__file__).resolve().parent
sys.path.insert(0, str(ICI))

RACINE = ICI.parents[1]
BUDGET = 267        # mm : valeur empirique, voir l'en-tête


def sonde(code):
    """Écrit une page-sonde : tous les blocs dans une colonne, mesurables."""
    from gabarit_fiche import CSS, charger_cours, h2, suite, cartouche, cloture

    mod = importlib.import_module(f"fiche_{code}")
    src = charger_cours(mod.COURS_DEFAUT)
    qr = mod.codes_qr(muet=True)
    blocs = mod.blocs(src, qr)

    morceaux = [
        f'<div class="sonde" data-i="{i}" data-titre="'
        f'{(p[0] + " " + p[1]) if p else ""}">{h2(*p) if p else html}</div>'
        for i, (p, html) in enumerate(blocs)
    ]
    # les elements de service, mesures eux aussi
    service = {
        "cartouche": cartouche("T", "C", "I", "x.jpg"),
        "suite": suite("01", "titre de continuation"),
        "cloture": cloture(qr[next(iter(qr))]),
    }
    morceaux += [f'<div class="sonde" data-i="s:{n}">{h}</div>'
                 for n, h in service.items()]

    # fonts.css en URL absolue : la sonde peut vivre hors du depot
    fonts = (RACINE / "assets" / "css" / "fonts.css").as_uri()
    html = (
        '<!DOCTYPE html><html lang="fr"><head><meta charset="UTF-8">\n'
        f'<link rel="stylesheet" href="{fonts}">\n<style>{CSS}\n'
        '  .feuille { height:auto !important; min-height:0 !important;\n'
        '             overflow:visible !important; }\n'
        '  .corps { display:block !important; }\n'
        # flow-root : la div contient les marges de ses enfants, donc sa
        # hauteur est bien celle qu'occupera le bloc dans la colonne.
        '  .sonde { display:flow-root; }\n'
        '</style></head><body>\n<div class="feuille"><div class="corps">'
        f'<div class="principal">\n{"".join(morceaux)}\n'
        '</div></div></div>\n</body></html>\n')

    out = Path(tempfile.gettempdir()) / f"sonde-fiche-{code}.html"
    out.write_text(html, encoding="utf-8", newline="\n")
    return out, len(blocs)


def main():
    if len(sys.argv) < 2:
        raise SystemExit(__doc__)
    code = sys.argv[1]
    pages = sys.argv[2] if len(sys.argv) > 2 else "0"
    budget = sys.argv[3] if len(sys.argv) > 3 else str(BUDGET)
    chemin, n = sonde(code)
    print(f"sonde : {n} blocs → {chemin}\n")
    subprocess.run(["node", str(ICI / "paginer.mjs"), chemin.as_uri(),
                    budget, pages], check=True)
    chemin.unlink(missing_ok=True)


if __name__ == "__main__":
    main()

File: _outils/fiches/paginer.mjs
// paginer.mjs — moitie « mesure » de l'outil de pagination des fiches.
// Lance par paginer.py, qui ecrit la page-sonde. Voir son en-tete pour le
// pourquoi. Ici : Chrome mesure chaque bloc, puis une programmation dynamique
// choisit les coupes.
// Le critere : on minimise la somme des carres des creux, ce qui
// repartit le vide au lieu de le concentrer sur une page.
import { spawn } from 'node:child_process';
import { setTimeout as pause } from 'node:timers/promises';

const BIN = 'C:/Program Files/Google/Chrome/Application/chrome.exe';
const URL = process.argv[2];
const BUDGET = Number(process.argv[3] || 267);   // mm utiles par page
const PAGES_VOULUES = Number(process.argv[4] || 0); // 0 = libre (mais pair)
const PORT = 9421;

const chrome = spawn(BIN, ['--headless=new', `--remote-debugging-port=${PORT}`,
  '--disable-gpu', '--hide-scrollbars', '--allow-file-access-from-files',
  '--window-size=900,1200',
  `--user-data-dir=${process.env.TEMP || '/tmp'}/chrome-paginer-fiches`,
  'about:blank'], { stdio: 'ignore' });

let wsUrl;
for (let i = 0; i < 80 && !wsUrl; i++) {
  try {
    const l = await (await fetch(`http://127.0.0.1:${PORT}/json/list`)).json();
    wsUrl = l.find(t => t.type === 'page')?.webSocketDebuggerUrl;
  } catch {}
  if (!wsUrl) await pause(250);
}
const ws = new WebSocket(wsUrl);
await new Promise(r => ws.addEventListener('open', r, { once: true }));
let id = 0; const att = new Map();
ws.addEventListener('message', e => {
  const m = JSON.parse(e.data);
  if (m.id && att.has(m.id)) { att.get(m.id)(m); att.delete(m.id); }
});
const envoi = (m, p = {}) => { const n = ++id;
  return new Promise(r => { att.set(n, r); ws.send(JSON.stringify({ id: n, method: m, params: p })); }); };

await envoi('Page.enable');
await envoi('Emulation.setEmulatedMedia', { media: 'print' });
await envoi('Page.navigate', { url: URL });
await pause(3500);

const r = await envoi('Runtime.evaluate', { returnByValue: true, expression: `
  (() => { const MM = 96 / 25.4;
    return [...document.querySelectorAll('.sonde')].map(d => {
      // display:flow-root -> la div contient les marges de ses enfants : sa
      // hauteur est bien celle qu'occupera le bloc dans la colonne.
      const e = d.firstElementChild || d;
      const h = d.getBoundingClientRect().height;
      return { i: d.dataset.i, titre: d.dataset.titre || '',
        etq: (e.querySelector('.etq')?.textContent || e.textContent).trim().slice(0,46).replace(/\\s+/g,' '),
        h: +(h / MM).toFixed(1) };
    });
  })()` });
const tous = r.result.result.value;
ws.close(); chrome.kill();

const svc = Object.fromEntries(tous.filter(b => b.i.startsWith('s:'))
  .map(b => [b.i.slice(2), b.h]));
const B = tous.filter(b => !b.i.startsWith('s:')).map(b => ({ ...b, i: +b.i }));

console.log(`\nblocs : ${B.length} · cartouche ${svc.cartouche} mm · suite ${svc.suite} mm · cloture ${svc.cloture} mm`);
console.log(`budget par page : ${BUDGET} mm\n`);
for (const b of B) console.log(`  ${String(b.i).padStart(3)}  ${String(b.h).padStart(6)} mm  ${b.titre || b.etq}`);

const total = B.reduce((s, b) => s + b.h, 0);
console.log(`\nsomme des blocs : ${total.toFixed(1)} mm`);

// un titre de partie (h2) ou de sous-partie (« A · … ») ne finit pas une page
const orphelin = b => b.titre !== '' || /^[A-H] · /.test(b.etq);

// cout d'une page [i, j) : surcharge fixe selon la position
function hauteur(i, j, n, N) {
  let h = B.slice(i, j).reduce((s, b) => s + b.h, 0);
  if (n === 0) h += svc.cartouche;
  else if (B[i].titre === '') h += svc.suite;   // page de continuation
  if (n === N - 1) h += svc.cloture;
  return h;
}

function paginer(N) {
  const n = B.length;
  const INF = 1e12;
  // dp[k][i] = cout minimal pour placer B[i..] en k pages
  const dp = Array.from({ length: N + 1 }, () => new Float64Array(n + 1).fill(INF));
  const ch = Array.from({ length: N + 1 }, () => new Int32Array(n + 1).fill(-1));
  dp[0][n] = 0;
  for (let k = 1; k <= N; k++) {
    for (let i = n - 1; i >= 0; i--) {
      for (let j = i + 1; j <= n; j++) {
        const h = hauteur(i, j, N - k, N);
        if (h > BUDGET) break;
        // 🔴 une page ne se termine jamais sur un titre (h2 « 03 … » ou
        // sous-partie « C · … ») : il serait seul en bas, séparé de ce qu'il
        // annonce. La première découpe de T1-C2 en laissait deux.
        if (j < n && orphelin(B[j - 1])) continue;
        const reste = dp[k - 1][j];
        if (reste >= INF) continue;
        const creux = BUDGET - h;
        const c = creux * creux + reste;
        if (c < dp[k][i]) { dp[k][i] = c; ch[k][i] = j; }
      }
    }
  }
  if (dp[N][0] >= INF) return null;
  const coupes = []; let i = 0;
  for (let k = N; k >= 1; k--) { const j = ch[k][i]; if (k > 1) coupes.push(j); i = j; }
  return { coupes, cout: dp[N][0] };
}

const candidats = PAGES_VOULUES ? [PAGES_VOULUES] : [10, 11, 12, 13, 14];
for (const N of candidats) {
  const res = paginer(N);
  if (!res) { console.log(`\n${N} pages : impossible dans ce budget`); continue; }
  const bornes = [0, ...res.coupes, B.length];
  const hs = [];
  for (let n = 0; n < N; n++) hs.push(hauteur(bornes[n], bornes[n + 1], n, N));
  const creux = hs.map(h => BUDGET - h);
  console.log(`\n${N} pages  COUPES = [${res.coupes.join(', ')}]`);
  console.log(`   hauteurs : ${hs.map(h => h.toFixed(0)).join(' · ')}`);
  console.log(`   creux    : ${creux.map(c => c.toFixed(0)).join(' · ')}  (cumulé ${creux.reduce((a, b) => a + b, 0).toFixed(0)} mm, max ${Math.max(...creux).toFixed(0)} mm)`);
}
